src/vm/readytoruninfo.cpp

```diff
@@ -890,6 +890,84 @@ PCODE ReadyToRunInfo::MethodIterator::GetMethodStartAddress()
     return ret;
 }
 
+BOOL ReadyToRunInfo::GenericMethodIterator::Next()
+{
+    m_current = m_enum.GetNext();
+    return !m_current.IsNull();
+}
+
+MethodDesc *ReadyToRunInfo::GenericMethodIterator::GetMethodDesc_NoRestore()
+{
+    _ASSERTE(!m_current.IsNull());
+
+    PCCOR_SIGNATURE pBlob = (PCCOR_SIGNATURE)m_current.GetBlob();
+    SigPointer sig(pBlob);
+    DWORD methodFlags = 0;
+    uint id = 0;
+    DWORD i = 0;
+    uint offset = 0;
+    uint val;
+    RID rid = 0;
+    DWORD numGenericArgs = 0;
+    PCCOR_SIGNATURE pSigNew = NULL;
+    DWORD cbSigNew = 0;
+    PCODE pEntryPoint = NULL;
+    MethodDesc *pMD = NULL;
+    HRESULT hr = S_OK;
+
+    // Skip the signature so we can get to the offset
+    IfFailGoto(sig.GetData(&methodFlags), Done);
+
+    if (methodFlags & ENCODE_METHOD_SIG_OwnerType)
+    {
+        IfFailGoto(sig.SkipExactlyOne(), Done);
+    }
+
+    _ASSERTE((methodFlags & ENCODE_METHOD_SIG_SlotInsteadOfToken) == 0);
+    _ASSERTE((methodFlags & ENCODE_METHOD_SIG_MemberRefToken) == 0);
+
+    IfFailGoto(sig.GetData(&rid), Done);
+    
+    if (methodFlags & ENCODE_METHOD_SIG_MethodInstantiation)
+    {
+        IfFailGoto(sig.GetData(&numGenericArgs), Done);
+    
+        for (i = 0; i < numGenericArgs; i++)
+        {
+            IfFailGoto(sig.SkipExactlyOne(), Done);
+        }
+    }
+
+    // Now that we have the size of the signature we can grab the offset and decode it
+    sig.GetSignature(&pSigNew, &cbSigNew);
+    offset = m_current.GetOffset() + (uint)(pSigNew - pBlob);
+
+    offset = m_pInfo->m_nativeReader.DecodeUnsigned(offset, &id);
```

Comment:
I'm not familiar enough with the native format of R2R images to speak to the soundness of this code.  It looks like this is traversing a variably sized method signature to find the offset for the `PCODE`.  If that is the case, can we simply add this entire code path as a "second phase" of the regular `MethodIterator`, e.g., after we iterate through the array of non-generic methods, instead of returning false @ `readytoruninfo.cpp:L833`, we simply start iterating the generic methods?

Reply:
It's possible, but I don't see a way currently to satisfy the other methods of `MethodIterator`, i.e. `GetMethodStartAddress` and `GetMethodDesc`, and it felt worse to have `MethodIterator` sometimes be able to satisfy those and sometimes not than to make a new class with limited functionality. I have to talk to Fadi and see if I'm missing something, I would actually prefer to have it all be part of MethodIterator since it would solve the other uses of `MethodIterator` (perfmap and ngen pdbs).